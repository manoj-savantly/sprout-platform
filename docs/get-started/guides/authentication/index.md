---
title: Authentication
description: Learn how to configure authentication for the Sprout UI and Server
---

Sprout supports several authentication mechanisms.  

## Basic  
Basic authentication is provided by default, and an initial admin user is created if you're using the default server.  
You may configure basic auth users to be created at startup using the application properties.  
If the user already exists, the configuration is skipped. This allows you to change a password after provisioning, without it being overwritten.  

Example configuration -  

```
sprout:
  security:
    authentication:
      basic:
        enable: true
        users:
        - username: admin
          password: changeme!
          email-address: example@savantly.net
```

## OAuth2  
You can provide provide the UI with an Implicit OAuth2 flow by configuring the properties.  

Example using keycloak -  
In this example, a realm called `my-realm` and a client called `webapp` have been created in keycloak.  
Implicit flow has also been enabled for the `webapp` client.  

```
sprout:
  security:
    authentication:
      oauth:
        clients:
        - name: keycloak
          display-name: Login with Keycloak Account
          authorization-url: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/auth
          token-url: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/token
          user-info-url: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/userinfo
          jwks-url: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/certs
          scope: openid profile email
          redirect-url: "http://localhost:8080/"
          clientId: webapp
          flow-type: IMPLICIT
spring:
  security:
    oauth2:
      resourceserver:
        jwt:
          jwk-set-uri: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/certs
```

OAuth can be further configured by providing your own implementations of certain beans.  
Each of these interfaces have default implementations that can be extended, or replaced.  

`OAuth2UserSynchronizer` executes after a successful authentication. It receieves an `OAuth2User` and should produce a `SproutUser`  
`OAuthConfigurer` configures the Spring `HttpSecurity`  
`OAuthUserMapper` maps an `OAuth2User` to a `SproutOAuthUser`  
`OAuth2UserService` receives an `OAuth2UserRequest` and produces an `OAuth2User`.  
`OidcUserService` receives an `OidcUserRequest` and produces an `OAuth2User`.  

## JWT
The backend server application uses standard Spring Authentication properties.  
Example of JWT configuration in the application.properties -  

```
spring:
  security:
    oauth2:
      resourceserver:
        jwt:
          jwk-set-uri: https://my-keycloak:9090/auth/realms/my-realm/protocol/openid-connect/certs
```

## Cookie
TODO: Improve Documentation  
After logging in, a session cookie is established to provide authentication when requesting static resources from a static site, such as the Sprout Web UI.  
