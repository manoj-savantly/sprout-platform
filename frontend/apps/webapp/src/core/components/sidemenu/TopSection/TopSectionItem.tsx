import { NavModelItem } from '@savantly/sprout-api';
import { MenuItem, SubMenu } from 'react-pro-sidebar';
import React, { FC } from 'react';
import { NavLink } from 'react-router-dom';
import { PrivateComponent } from '../../PrivateComponent/PrivateComponent';
import { Icon } from '@sprout-platform/ui';
import useDevice from '../../../hooks/useDevice';

export interface Props {
  link: NavModelItem;
  onClick?: () => void;
  firstChild?: boolean;
}

const TopSectionItem: FC<Props> = (props) => {
  const { link, onClick, firstChild } = props;
  const { isMobile } = useDevice();

  const leafNode = (
    <MenuItem icon={<Icon name={link.icon || ('cube' as any)} size="1x" />}>
      {link.onClick ? (
        <span onClick={link.onClick}>{link.text}</span>
      ) : (
        <NavLink to={link.url || '/'} onClick={onClick}>
          {link.text}
        </NavLink>
      )}
    </MenuItem>
  );

  const hasChildren = !!link.children?.length;

  const linkTag = (
    <>
      {!hasChildren ? (
        isMobile || !firstChild ? (
          leafNode
        ) : (
          <SubMenu
            title={link.text}
            icon={<Icon name={link.icon || ('cube' as any)} size="1x" />}
            firstchild={firstChild}
          >
            {leafNode}
          </SubMenu>
        )
      ) : (
        <SubMenu
          title={link.text}
          icon={<Icon name={link.icon || ('cube' as any)} size="1x" />}
          firstchild={firstChild}
        >
          {(link.children || []).map((child) => (
            <TopSectionItem key={child.id} link={child} />
          ))}
        </SubMenu>
      )}
    </>
  );

  return link.authority ? <PrivateComponent hasAnyAuthority={[link.authority]}>{linkTag}</PrivateComponent> : linkTag;
};

export default TopSectionItem;
