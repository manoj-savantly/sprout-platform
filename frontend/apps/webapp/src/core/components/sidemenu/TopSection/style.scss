.TopSection__mobileMenuIcon {
  position: absolute;
  top: 10px;
  left: 10px;
  z-index: 1000;
}

.TopSection__logoContainer {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 50px;
}