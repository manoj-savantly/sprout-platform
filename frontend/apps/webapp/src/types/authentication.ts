import { User } from '@savantly/sprout-api';

export interface AuthenticationState {
  loading: boolean;
  isAuthenticated: boolean;
  loginError: boolean; // Errors returned from server side
  errorMessage?: string; // Errors returned from server side
  sessionHasBeenFetched: boolean;
  sessionFetchFailed: boolean;
  logoutUrl?: string;
  user: User;
  showLogin: boolean;
}

export interface AuthenticationUpdate {
  errorMessage?: string;
  accessToken?: string;
}

export interface OAuthClientConfig {
  name: string;
  autoLogin?: boolean;
  displayName: string;
  issuerUri: string;
  clientId: string;
  scope: string;
  authorizationUrl: string;
  redirectUrl: string;
  tokenUrl: string;
  userInfoUrl: string;
}