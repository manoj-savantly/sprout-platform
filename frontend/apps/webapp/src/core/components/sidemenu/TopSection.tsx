import _ from 'lodash';
import React, { FC } from 'react';
import { useSelector } from 'react-redux';
import { StoreState } from '../../../types';
import { PrivateComponent } from '../PrivateComponent/PrivateComponent';
import TopSectionItem from './TopSectionItem';

const TopSection: FC<any> = () => {
  const navTree = useSelector((state: StoreState) => state.navTree.items);
  const mainLinks = _.filter(navTree, (item) => !item.hideFromMenu).sort(
    (a, b) => (a?.position || 0) - (b?.position || 0)
  );
  console.log(mainLinks);

  return (
    <div className="sidemenu__top">
      {/* <TopSectionItem link={searchLink} onClick={onOpenSearch} /> */}
      {mainLinks.map((link, index) => {
        return (
          (link.authority && (
            <PrivateComponent hasAnyAuthority={[link.authority]} key={`pc-${link.id}-${index}`} >
              <TopSectionItem link={link} key={`${link.id}-${index}`} />
            </PrivateComponent>
          )) || <TopSectionItem link={link} key={`${link.id}-${index}`} />
        );
      })}
    </div>
  );
};

export default TopSection;
