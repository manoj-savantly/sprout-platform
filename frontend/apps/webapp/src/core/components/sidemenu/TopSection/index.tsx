import React, { FC, useState, useEffect } from 'react';
import { NavLink } from 'react-router-dom';
import { ProSidebar, Menu, MenuItem, SubMenu, SidebarContent, SidebarHeader } from 'react-pro-sidebar';
import { useSelector } from 'react-redux';
import { StoreState } from '../../../../types';
import { filter } from 'lodash';
import TopSectionItemNew from './TopSectionItemNew';
import { HamburgerIcon } from '@chakra-ui/icons';

import 'react-pro-sidebar/dist/css/styles.css';
import './style.scss';
import { Branding } from '../../Branding/Branding';
import useDevice from '../../../hooks/useDevice';

const TopSectionNew: FC<any> = () => {
  const navTree = useSelector((state: StoreState) => state.navTree.items);
  const mainLinks = filter(navTree, (item) => !item.hideFromMenu).sort(
    (a, b) => (a?.position || 0) - (b?.position || 0)
  );
  const { isMobile } = useDevice();
  const [toggled, onToggle] = useState(false);
  const [collapsed, setCollapsed] = useState(!isMobile);
  useEffect(() => {
    if (isMobile) {
      return setCollapsed(false);
    }
    setCollapsed(true);
  }, [isMobile]);

  return (
    <>
      {isMobile ? (
        <div className="TopSection__mobileMenuIcon">
          <HamburgerIcon w={20} h={20} onClick={() => onToggle(true)} />
        </div>
      ) : null}
      <ProSidebar collapsed={collapsed} breakPoint="md" toggled={toggled} onToggle={onToggle}>
        <SidebarHeader>
          <span className="TopSection__logoContainer">
            <NavLink to={'/'}>
              <Branding.MenuLogo width={30} />
            </NavLink>
          </span>
        </SidebarHeader>
        <SidebarContent>
          <Menu iconShape="circle" popperArrow={true}>
            {mainLinks.map((link) => {
              return <TopSectionItemNew key={link.id} link={link} firstChild={true} />;
            })}
          </Menu>
        </SidebarContent>
      </ProSidebar>
    </>
  );
};

export default TopSectionNew;
