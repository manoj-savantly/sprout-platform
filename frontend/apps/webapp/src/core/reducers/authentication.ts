import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { User } from '@savantly/sprout-api';
import axios, { AxiosRequestConfig } from 'axios';
import { ACCESS_TOKEN_STORAGE_KEY, SERVER_API_URL } from '../../config/constants';
import { AuthenticationState, AuthenticationUpdate } from '../../types';
import store from '../store';

export const initialAuthenticationState: AuthenticationState = {
  loading: false,
  isAuthenticated: store.get(ACCESS_TOKEN_STORAGE_KEY) ? true : false,
  loginError: false, // Errors returned from server side
  user: {
    authorities: []
  } as User,
  errorMessage: undefined, // Errors returned from server side
  sessionHasBeenFetched: false,
  sessionFetchFailed: false,
  logoutUrl: undefined,
  showLogin: false
};

export const getSession = createAsyncThunk('authentication/getSession', async (arg, thunkAPI) => {
  const token = store.get(ACCESS_TOKEN_STORAGE_KEY);
  const config: AxiosRequestConfig = {
    headers: {}
  };
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  } else {
    console.log('sending unauthenticated request', config);
  }
  const response = await axios.get(`${SERVER_API_URL}/api/account`, config);
  console.log(response);
  if (response.data) {
    return response.data;
  } else {
    store.delete(ACCESS_TOKEN_STORAGE_KEY);
    return thunkAPI.rejectWithValue(response);
  }
});

const authenticationSlice = createSlice({
  name: 'authentication',
  initialState: initialAuthenticationState,
  reducers: {
    showLogin: (state): AuthenticationState => {
      return {
        ...state,
        showLogin: true
      };
    },
    login: (state, action: PayloadAction<AuthenticationUpdate>): AuthenticationState => {
      const isAuthenticated = action.payload.accessToken ? true : false;
      if (isAuthenticated) {
        store.set(ACCESS_TOKEN_STORAGE_KEY, action.payload.accessToken as string);
      }
      return {
        ...state,
        isAuthenticated,
        errorMessage: action.payload.errorMessage,
        loginError: !isAuthenticated,
        showLogin: !isAuthenticated,
        sessionHasBeenFetched: false,
        sessionFetchFailed: false
      };
    },
    logout: (state): AuthenticationState => {
      if (store.get(ACCESS_TOKEN_STORAGE_KEY)) {
        store.delete(ACCESS_TOKEN_STORAGE_KEY);
      }
      return {
        ...state,
        user: {
          authorities: []
        },
        errorMessage: undefined,
        isAuthenticated: false,
        loading: false,
        loginError: false,
        logoutUrl: undefined,
        showLogin: false
      };
    }
  },
  extraReducers: (builder) => {
    builder.addCase(
      getSession.fulfilled,
      (state, action): AuthenticationState => {
        console.log(action);
        return {
          ...state,
          user: {
            ...state.user,
            ...action.payload
          },
          isAuthenticated: true,
          sessionHasBeenFetched: true,
          sessionFetchFailed: false,
          showLogin: false
        };
      }
    );
    builder.addCase(
      getSession.rejected,
      (state, action): AuthenticationState => {
        console.error(action);
        const authError = !!action.error.message && action.error.message.indexOf('401') > -1;
        return {
          ...state,
          user: {
            ...state.user
          },
          isAuthenticated: false,
          sessionHasBeenFetched: false,
          sessionFetchFailed: true,
          showLogin: authError
        };
      }
    );
  }
});

export const { login, logout, showLogin } = authenticationSlice.actions;

export const authenticationReducer = authenticationSlice.reducer;
