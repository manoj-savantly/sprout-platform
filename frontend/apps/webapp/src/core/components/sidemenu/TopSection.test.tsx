import { NavModelItem } from '@savantly/sprout-api';
import { render } from 'enzyme';
import moxios from 'moxios';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { initialAuthenticationState } from '../../reducers/authentication';
import TopSection from './TopSection';

const mockNavItems: NavModelItem[] = [
  {
    text: 'open'
  },
  {
    text: 'private',
    authority: 'PRIVATE'
  }
];

const mockStore = configureStore([]);

describe('TopSection when session fetched but no authority', () => {
  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  const store = mockStore({
    authentication: initialAuthenticationState,
    navTree: {
      items: mockNavItems
    }
  });

  it('renders 1 with no authority requirement', () => {
    const component = render(
      <BrowserRouter>
        <Provider store={store}>
          <TopSection />
        </Provider>
      </BrowserRouter>
    );
    const ss = component; //.find('.side-menu-item');
    expect(ss.children().length).toBe(1);
    expect(ss).toMatchSnapshot();
  });

});
