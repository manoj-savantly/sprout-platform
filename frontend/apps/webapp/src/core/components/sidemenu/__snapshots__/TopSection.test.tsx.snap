// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopSection when session fetched but no authority renders 1 with no authority requirement 1`] = `
<div
  class="sidemenu__top"
>
  <div
    class="sidemenu-item dropdown justify-content-center css-175dj8u"
  >
    <a
      class="sidemenu-link"
    >
      <span
        class="icon-circle sidemenu-icon"
      >
        <div
          class="css-1vzus6i-Icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-cube fa-w-16 fa-1x css-1u3o4ib"
            data-icon="cube"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M239.1 6.3l-208 78c-18.7 7-31.1 25-31.1 45v225.1c0 18.2 10.3 34.8 26.5 42.9l208 104c13.5 6.8 29.4 6.8 42.9 0l208-104c16.3-8.1 26.5-24.8 26.5-42.9V129.3c0-20-12.4-37.9-31.1-44.9l-208-78C262 2.2 250 2.2 239.1 6.3zM256 68.4l192 72v1.1l-192 78-192-78v-1.1l192-72zm32 356V275.5l160-65v133.9l-160 80z"
              fill="currentColor"
            />
          </svg>
        </div>
      </span>
    </a>
    <ul
      class="dropdown-menu dropdown-menu--sidemenu"
      role="menu"
    >
      <li
        class="side-menu-header"
      >
        <a
          aria-current="page"
          class="side-menu-header-link active"
          href="#"
        >
          <span
            class="sidemenu-item-text"
          >
            open
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
